source/common/http/hash_policy.cc

```diff
@@ -39,14 +42,31 @@ class HeaderHashMethod : public HashMethodImplBase {
                                     const StreamInfo::FilterStateSharedPtr) const override {
     absl::optional<uint64_t> hash;
 
-    // TODO(mattklein123): Potentially hash on all headers.
     const auto header = headers.get(header_name_);
     if (!header.empty()) {
       if (regex_rewrite_ != nullptr) {
-        hash = HashUtil::xxHash64(regex_rewrite_->replaceAll(header[0]->value().getStringView(),
+        std::vector<std::string> header_values;
```

Comment:
Your implementation looks correct to me. But WDYT of de-duping the 2 10-line blocks of code below. I'd do it by always using `vector<string_view>`, but also declare `vector<string> buffers` for backing-store for regexes. Then after initially populating the un-transformed header_values, you'd add:
```
if (regex_rewrite_ != nullptr) {
  for (absl::string_view& val : header_values) {
    buffers_.push_back(regex_rewrite_->replaceAll(val, regex_rewrite_substitution_));
    val =   buffers_.back();
  }
}

Then the hasher would always use `vector<absl::string_view>`, and you'd de-dup the hashing and sorting logic.

Just a thought. If you like it better you can switch to that or I'm OK leaving it as is. Either way you need to test all corner-cases with and without regexes, per comments below.